servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublishAndSubscribeOnPublishers.java

```diff
@@ -88,26 +94,26 @@ public void handleSubscribe(final Subscriber<? super T> subscriber,
      */
     private static final class SubscribeOn<T> extends TaskBasedAsyncPublisherOperator<T> {
 
-        SubscribeOn(final Publisher<T> original, final Executor executor) {
-            super(original, executor);
-        }
-
-        @Override
-        public Subscriber<? super T> apply(Subscriber<? super T> subscriber) {
-            return new OffloadedSubscriptionSubscriber<>(subscriber, executor());
+        SubscribeOn(final Publisher<T> original,
+                    final Supplier<? extends BooleanSupplier> shouldOffloadSupplier, final Executor executor) {
+            super(original, shouldOffloadSupplier, executor);
         }
 
         @Override
         public void handleSubscribe(final Subscriber<? super T> subscriber,
                                     final AsyncContextMap contextMap, final AsyncContextProvider contextProvider) {
             try {
-                // re-wrap the subscriber so that async context is restored during offloading.
-                Subscriber<? super T> wrapped =
-                        contextProvider.wrapSubscription(subscriber, contextMap);
-
-                // offload the remainder of subscribe()
-                LOGGER.trace("Offloading Publisher subscribe() on {}", executor());
-                executor().execute(() -> super.handleSubscribe(wrapped, contextMap, contextProvider));
+                BooleanSupplier shouldOffload = shouldOffloadSupplier();
+                final Subscriber<? super T> upstreamSubscriber =
+                        new OffloadedSubscriptionSubscriber<>(subscriber, shouldOffload, executor());
+
+                if (shouldOffload.getAsBoolean()) {
```

Comment:
General comment for all async sources: I see that inside `publishOn` operator we log and ignore exceptions from the `BooleanSupplier`, but `subscribeOn` does not ignore them and terminates the `Subscriber`. Can we be consistent?

Reply:
This may be moot if we remove the `Supplier<? extends BooleanSupplier>` layer as in the publishOn case it was only catching the possible exception from the `Supplier.get()`. Sufficient?

Reply:
I was referring to this method:
https://github.com/apple/servicetalk/blob/603b2dc4f2136c1acd153d21afb94d960fb137c2/servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/TaskBasedAsyncCompletableOperator.java#L348-L355

It's used to invoke the `BooleanSupplier`.

Reply:
Gotcha. I agree that it should cancel and the existing comments in the methods calling `shouldOffload()` indicate that is the required policy.

Reply:
This was improved in #1723, thank you @bondolo!